File: 695-max-area-of-island/max-area-of-island.cpp
class Solution {
public:
    void dfs(int row, int col, vector<vector<int>>& grid, vector<vector<bool>>& visited, int& area) {
        int rows = grid.size();
        int cols = grid[0].size();

        if (row < 0 || row >= rows || col < 0 || col >= cols)
            return;
        if(grid[row][col] == 0 || visited[row][col])
            return;
      
        area++;
        visited[row][col] = true;

        dfs(row - 1, col, grid, visited, area);
        dfs(row + 1, col, grid, visited, area);
        dfs(row, col - 1, grid, visited, area);
        dfs(row, col + 1, grid, visited, area);
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();
        int maxArea = 0;

        vector<vector<bool>> visited(rows, vector<bool>(cols, false));
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] == 1 && !visited[i][j]) {
                    int area = 0;
                    dfs(i, j, grid, visited, area);
                    maxArea = max(area, maxArea);
                }
            }
        }
        return maxArea;
    }
};

File: 695-max-area-of-island/max-area-of-island.java
class Solution {
    public int rows, cols;
    
    public int DFS(int i, int j, int[][] grid, boolean[][] visited) {
        if(i < 0 || i >= rows || j < 0 || j >= cols) 
            return 0;

        if(grid[i][j] == 0 || visited[i][j])
            return 0;
        
        visited[i][j] = true;
        
        return 1 
            + DFS(i + 1, j, grid, visited)
            + DFS(i - 1, j, grid, visited)
            + DFS(i, j + 1, grid, visited)
            + DFS(i, j - 1, grid, visited);
    }

    public int maxAreaOfIsland(int[][] grid) {
        this.rows = grid.length;
        this.cols = grid[0].length;

        boolean[][] visited = new boolean[rows][cols];
        int maxArea = 0;
        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] == 1 && !visited[i][j]) {
                    int area = DFS(i, j, grid, visited);
                    maxArea = Math.max(maxArea, area);
                }
            }
        }
        return maxArea;
    }
}